urle381b6cba7_IkarosKappler_ikrs_http_tgz-pJ8-ikrs_json_rpc_JSONRPCHandlerJ8/src/ikrs/yuccasrv/socketmngr/ServerSocketThread.java

```diff
@@ -0,0 +1,468 @@
+
+	this.bindAddress = bindAddress;
+	this.bindPort = bindPort;
+	this.bindSettings = bindSettings;
+	
+	this.observer = observer;
+
+	int backlog = DEFAULT_BACKLOG;
+	if( bindSettings.get(Constants.CONFIG_SERVER_BACKLOG) != null )
+	    backlog = bindSettings.get(Constants.CONFIG_SERVER_BACKLOG).getInt( DEFAULT_BACKLOG );
+
+	boolean useSSL = false;
+	if( bindSettings.get(Constants.CONFIG_SERVER_SSL) != null )
+	    useSSL = bindSettings.get(Constants.CONFIG_SERVER_SSL).getBoolean();
+
+
+
+	if( useSSL ) {
+
+	    try {
+		this.serverSocket = createSSLServerSocket( bindPort,
+							   backlog,
+							   bindAddress,
+							   this.bindSettings 
+							   );
+	    } catch( NoSuchAlgorithmException e ) {
+
+		throw new GeneralSecurityException( "Internal error: Cannot create SSLServerSocket. Failed to load encryption algorithm.", e );
+
+	    }
+
+	} else {
+	    
+	    this.serverSocket = new ServerSocket( bindPort,
+						  backlog,
+						  bindAddress 
+						  );
+	}
+
+	this.uuid = UUID.randomUUID();
+
+    }
+
+    private SSLServerSocket createSSLServerSocket( int bindPort,
+						   int backlog,
+						   InetAddress bindAddress,
+						   Environment<String,BasicType> bindSettings ) 
+	throws IOException,
+	       FileNotFoundException,
+	       KeyStoreException,
+	       KeyManagementException,
+	       NoSuchAlgorithmException,
+	       CertificateException,
+	       UnrecoverableKeyException {
+
+
+	BasicType wrp_keyStoreName    = getSSLConfigPropertyValue( bindSettings, Constants.CONFIG_SERVER_SSL_KEYSTORE );
+	BasicType wrp_keyStorePass    = getSSLConfigPropertyValue( bindSettings, Constants.CONFIG_SERVER_SSL_KEYSTOREPASSWORD );
+	BasicType wrp_keyStoreType    = getSSLConfigPropertyValue( bindSettings, Constants.CONFIG_SERVER_SSL_KEYSTORETYPE );
+
+	BasicType wrp_trustStoreName  = getSSLConfigPropertyValue( bindSettings, Constants.CONFIG_SERVER_SSL_TRUSTSTORE );
+	BasicType wrp_trustStorePass  = getSSLConfigPropertyValue( bindSettings, Constants.CONFIG_SERVER_SSL_TRUSTSTOREPASSWORD );
+	BasicType wrp_trustStoreType  = getSSLConfigPropertyValue( bindSettings, Constants.CONFIG_SERVER_SSL_TRUSTSTORETYPE );
+
+	BasicType wrp_wantClientAuth  = getSSLConfigPropertyValue( bindSettings, Constants.CONFIG_SERVER_SSL_WANTCLIENTAUTH );
+	BasicType wrp_needClientAuth  = getSSLConfigPropertyValue( bindSettings, Constants.CONFIG_SERVER_SSL_NEEDCLIENTAUTH );
+
+
+	KeyStore keyStore = null;
+	if( wrp_keyStoreName != null && wrp_keyStoreName.getString().length() != 0 ) {
+
+	    String keyStoreName = Yucca.processCustomizedFilePath( wrp_keyStoreName.getString() );
+	    logger.log( Level.INFO,
+			getClass().getName() + ".createSSLServerSocket(...)",
+			"Loading key store '" + keyStoreName + "' ... " );
+	    keyStore = loadKeyStore( keyStoreName,
+				     (wrp_keyStorePass==null?new char[0]:wrp_keyStorePass.getString().toCharArray()),
+				     (wrp_keyStoreType==null?null:wrp_keyStoreType.getString())
+				     );
+	    
+	}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 151:162, 166:177.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.